src/tools/crawling_tools.py

```
@@ -390,43 +423,90 @@ async def crawl(ctx: Context, url: str, max_depth: int = 3, max_concurrent: int
+        if "QdrantException" in str(type(e)):
+            error = error_handler.handle_database_error(e, "sources", "scroll")
+        else:
+            error = error_handler.create_error(
+                e,
+                category=ErrorCategory.INTERNAL,
+                error_code="SOURCES_FETCH_FAILED"
+            )
+        return await error_handler.log_and_report_error(ctx, error)
 
 @mcp.tool()
 async def rag_query(ctx: Context, query: str, source: str = None, match_count: int = 5) -> str:
```

Ruff (0.12.2)

`rag_query` is too complex (11 > 10)

(C901)